### PR TITLE
Centralize admin implementations and action execution

### PR DESCRIPTION
**Pull request by kimchy**
This change removes the multiple implementations of different admin interfaces and centralizes it with AbstractClient. It also makes sure _all_ executions of actions now go through a single AbstractClient#execute method, taking care of copying headers and wrapping listener.
This also has the side benefit of removing all the code around differnet possible clients, and removes quite a bit of code (most of the + code is actually removal of generics and such).

This change also changes how TransportClient is constructed, requiring a Builder to create it, its a breaking change and its noted in the migration guide.

Yea another step towards simplifying the action infra and making it simpler...

### CODE REVIEW

File: src/test/java/org/elasticsearch/client/transport/InternalTransportClientTests.java [hunk 1]
```diff
@@ -1,342 +0,0 @@
...
-import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicInteger;
-import java.util.concurrent.atomic.AtomicReference;
-
-import static org.hamcrest.CoreMatchers.*;
-import static org.hamcrest.Matchers.lessThanOrEqualTo;
-import static org.hamcrest.Matchers.notNullValue;
-
-public class InternalTransportClientTests extends ElasticsearchTestCase {
```
Comment (javanna) [inline src/test/java/org/elasticsearch/client/transport/InternalTransportClientTests.java]: Although this test tested a class that is now gone, I think the retry mechanism is still around and it should maybe be replaced with a corresponding new test. The issue that originated the test is #6829 . Thoughts?

File: src/main/java/org/elasticsearch/action/Action.java [hunk 2]
```diff
@@ -24,7 +24,7 @@
 /**
  * Base action. Supports building the <code>Request</code> through a <code>RequestBuilder</code>.
  */
-public abstract class Action<Request extends ActionRequest, Response extends ActionResponse, RequestBuilder extends ActionRequestBuilder<Request, Response, RequestBuilder, Client>, Client extends ElasticsearchClient>
+public abstract class Action<Request extends ActionRequest, Response extends ActionResponse, RequestBuilder extends ActionRequestBuilder<Request, Response, RequestBuilder>>
         extends GenericAction<Request, Response> {
```
Comment (javanna) [inline src/main/java/org/elasticsearch/action/Action.java]: while you are at it, can we remove the `Request` type definition from `GenericAction`, it seems unused ?

File: src/test/java/org/elasticsearch/client/transport/InternalTransportClientTests.java (hunk 1 above)
Comment (kimchy) [inline src/test/java/org/elasticsearch/client/transport/InternalTransportClientTests.java]: I looked at it, and got the second that `TransportClientNodesServiceTests` already tests it, and the fact that we centralized the execution, we don't need the double tests?

File: src/main/java/org/elasticsearch/client/support/AbstractClient.java [hunk 3]
```diff
@@ -548,6 +782,838 @@ public void fieldStats(FieldStatsRequest request, ActionListener<FieldStatsRespo
...
+
+        @Override
+        public <Request extends ActionRequest, Response extends ActionResponse, RequestBuilder extends ActionRequestBuilder<Request, Response, RequestBuilder>> RequestBuilder prepareExecute(Action<Request, Response, RequestBuilder> action) {
+            return client.prepareExecute(action);
+        }
+
+        @Override
+        public ThreadPool threadPool() {
+            return client.threadPool();
+        }
```
Comment (javanna) [inline src/main/java/org/elasticsearch/client/support/AbstractClient.java]: maybe the 4 methods above can be shared, given that they seem repeated across the different clients here?

**Comment by javanna** [comment]
very nice @kimchy ! left a couple of minor comments

File: src/test/java/org/elasticsearch/client/transport/InternalTransportClientTests.java (hunk 1 above)
Comment (javanna) [inline src/test/java/org/elasticsearch/client/transport/InternalTransportClientTests.java]: You are right, I totally missed that other test, great!

File: src/main/java/org/elasticsearch/action/Action.java (hunk 2 above)
Comment (kimchy) [inline src/main/java/org/elasticsearch/action/Action.java]: We might need it in the future (for another refactoring), if we don't, I would just remove it in a different change

File: src/main/java/org/elasticsearch/client/support/AbstractClient.java (hunk 3 above)
Comment (kimchy) [inline src/main/java/org/elasticsearch/client/support/AbstractClient.java]: Do you mean we create another base class? felt win/loss was towards this way

**Comment by kimchy** [comment]
@javanna answered them, thanks for the review!

**Comment by s1monw** [comment]
this one is huge +1 lets get it in

**Pull request merged**